[Custom Fields] Handle saving changes

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12459 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR includes the following changes:
- Adds logic to return results from the editor to the list screen.
- Show a save button on the list screen when there are pending changes.
- Update the `meta_data` array of the parent item when saving changes.

### Steps to reproduce
1. Open wp-admin and add some "Custom Fields" to one order.
2. Open the app.
3. Open the order.
4. Tap on "View custom fields"
5. Open some of the fields, and make some changes.
6. Notice the Save button is displayed.
7. Tap on Back.
8. Notice the discard changes dialog.
9. Tap on Cancel.
10. Tap on Save.
11. Confirm the saving worked fine.

### Testing information
- It would be great to test the happy and unhappy paths.
- Tested using the Android emulator running Android 13.

### Images/gif
https://github.com/user-attachments/assets/adf132c1-c33f-444e-a83a-a6fbd2b286cf


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Handle returning result to the list screen

hichamboushaba committed:
Add logic to reflect pending changes on the UI

hichamboushaba committed:
Pass the current value directly to the editor

This is needed to avoid displaying the DB value when we have some pending changes.

hichamboushaba committed:
Improve the logic of setting the pending changes

hichamboushaba committed:
Show Save button when there are changes

hichamboushaba committed:
Rename property

hichamboushaba committed:
Add a new property to show a dialog when saving changes

hichamboushaba committed:
Add repository function to handle saving changes

hichamboushaba committed:
Handle saving changes in the ViewModel

hichamboushaba committed:
Update messages shown after error and success

hichamboushaba committed:
Show a discard changes dialog when we have some pending changes

dangermattic commented:
<!--
  0 Errors
  1 Warning: Class <code>DiscardChangesDial...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>DiscardChangesDialogState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Use a different key for the field draft

`customField` matches the argument name, and this would create issues with the creation flow, as we'll initialize the SavedState value with `null` where it shouldn't.

hichamboushaba committed:
Fix editor tests

hichamboushaba committed:
Separate update and insert functions

hichamboushaba committed:
Add new unit tests

hichamboushaba committed:
Fix detekt issues

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorViewModel.kt:
```diff
@@ -3,97 +3,61 @@ package com.woocommerce.android.ui.customfields.editor
-        initialValue = null,
-        key = "customField",
-        clazz = CustomFieldUiModel::class.java
+        initialValue = navArgs.customField ?: CustomFieldUiModel("", ""),
```
Instead of passing just the `id`, we'll need to pass the whole custom field, this is because the value could be modified and not saved yet, so if we pass just the `id`, the editor will show the DB value which is different from what the user was seeing in the list.

hichamboushaba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/customfields/editor/CustomFieldsEditorViewModelTest.kt:
```diff
@@ -1,43 +1,59 @@
+    private lateinit var textUtilsStaticMock: MockedStatic<TextUtils>
+
+    @Before
+    fun prepare() {
+        /** [HtmlUtils.fastStripHtml] uses internally [TextUtils.isEmpty], so we need to mock it */
+        textUtilsStaticMock = mockStatic(TextUtils::class.java)
+        whenever(TextUtils.isEmpty(any())).thenAnswer { invocation ->
+            invocation.getArgument<String>(0).isEmpty()
+        }
     }
 
-    private lateinit var viewModel: CustomFieldsEditorViewModel
+    @After
+    fun tearDown() {
+        textUtilsStaticMock.close()
+    }
```
I don't like this mocking here, but we need it right now because the Wordpress utils library uses the `TextUtils.isEmpty` internally.

If I can find time, I might update the library to remove this usage, then remove this mock here later.

irfano commented:
I haven't reviewed the code yet, but I tested it and found an issue. When I edit the field a second time, the changes are ignored. Let me know if you’re unable to reproduce the issue. I can continue reviewing the code once this is fixed.

[bug.webm](https://github.com/user-attachments/assets/6b2ce010-a58e-4809-a28e-792530ead7b9)

hichamboushaba committed:
Make sure to keep only one entry when editing a field

hichamboushaba commented:
> I haven't reviewed the code yet, but I tested it and found an issue. When I edit the field a second time, the changes are ignored. Let me know if you’re unable to reproduce the issue. I can continue reviewing the code once this is fixed.

Thanks @irfano, nice catch, actually I thought I already fixed this issue, but I think I didn't commit the changes while refactoring some parts 🤦.
Anyway, since I already knew the cause, the fix was straightforward, I also added a unit test for this case, all done in the commit 63e98715edeca725af7385bf3759ec051f847e5c

Ready for code review now

irfano commented:
One more small issue: The behavior of the back button and system back navigation should be the same on the field editing screen. System back navigation doesn't display a dialog.

[Screen_recording_20240910_180523.webm](https://github.com/user-attachments/assets/bb39bed6-486b-4003-b119-ff9ae80ace01)

irfano commented:
**Suggestion:** When I tap just on the value ("irfan" in the screenshot), the detail screen doesn't open. I guess the reason is that values can be clickable as a view. If the value is not clickable, we could set the view "clickable=false" for the view. I believe this should resolve it.

![image](https://github.com/user-attachments/assets/6ee51cd9-79e8-4ea8-ae5d-e2651b545a4a)

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/customfields/list/CustomFieldsFragment.kt:
```diff
@@ -52,11 +55,20 @@ class CustomFieldsFragment : BaseFragment() {
 
+    private fun handleResults() {
+        handleResult<CustomFieldUiModel>(CustomFieldsEditorFragment.RESULT_KEY) { result ->
+            if (result.id == null) {
+                viewModel.onCustomFieldInserted(result)
+            } else {
+                viewModel.onCustomFieldUpdated(result)
+            }
```
minor suggestion: We can pass the `result` directly to the view model, allowing the view model to decide whether to insert or update. It's a small logic, but it might fit better in the view model.

hichamboushaba replied:
Actually I decided to do it this way because the List ViewModel doesn't have to care about how the result is passed, the editor could choose to use different keys for passing insertions and additions.

But anyway, I see this goes this way or the other, so let's keep it this way :).

irfano reviewed (changes requested):
My review is complete. Everything works as expected, except for the points I mentioned in my comments. I can review it again once those are fixed.

hichamboushaba committed:
Attach a back handler to pass back button clicks to the viewmodel

hichamboushaba commented:
> One more small issue: The behavior of the back button and system back navigation should be the same on the field editing screen. System back navigation doesn't display a dialog.

Good catch, this should be fixed by 4e700bb

> Suggestion: When I tap just on the value ("irfan" in the screenshot), the detail screen doesn't open. I guess the reason is that values can be clickable as a view. If the value is not clickable, we could set the view "clickable=false" for the view. I believe this should resolve it.

Yes, this is a known issue, I'm planing to address it on a separate PR to keep the PRs more focused.

irfano reviewed (approved):
Thanks for the fix. LGTM! 👍🏻
